### PR TITLE
Fix plain clang on windows

### PR DESCRIPTION
**Pull request by dcbaker**
on Windows clang uses link.exe or lld-link.exe. It cannot (at least in my testing) use ld.lld (even though it installs it) because it has hardcoded arguments that ld.lld doesn't understand (-nologo, for example).

To fix this I've pulled `-Wl,` and similar out of the linkers as hardcoded values, and instead passed the value into the linkers at creation time. This has some additional nice affects like making the D and cuda compilers a little cleaner

### CODE REVIEW

File: mesonbuild/linkers.py [hunk 1]
```diff
@@ -801,18 +811,21 @@ class MSVCDynamicLinker(VisualStudioLikeLinkerMixin, DynamicLinker):
...
 
 class ClangClDynamicLinker(VisualStudioLikeLinkerMixin, DynamicLinker):
 
     """Clang's lld-link.exe."""
 
-    def __init__(self, for_machine: mesonlib.MachineChoice,
-                 *, version: str = 'unknown version'):
-        super().__init__(['lld-link.exe'], for_machine, 'lld-link',
+    def __init__(self, for_machine: mesonlib.MachineChoice, *,
+                 exelist: typing.List[str] = ['lld-link.exe'], prefix: str = '',
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [B006] Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them. ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4487208))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/cuda.py [hunk 2]
```diff
@@ -162,17 +165,8 @@ def has_header_symbol(self, hname, symbol, prefix, env, extra_args=None, depende
...
-        """
-        cooked = []  # type: typing.List[str]
-        for arg in args:
-            if arg.startswith('-Wl,'):
-                arg = arg.replace('-Wl,', '-Xlinker=', 1)
-            arg = arg.replace(' ', '\\')
-            cooked.append(arg)
-        return cooked
+        """Fixup arguments."""
+        return [arg.replace( ' ', '\\') for a in args]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/cuda.py]: __[Flake8]__ [E201] whitespace after '(' ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4487209))

<sub>posted by Sider</sub>

Comment (ignatenkobrain) [inline mesonbuild/compilers/cuda.py]: __[Flake8]__ [F821] undefined name 'arg' ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4487210))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 3]
```diff
@@ -801,18 +811,21 @@ class MSVCDynamicLinker(VisualStudioLikeLinkerMixin, DynamicLinker):
 
     """Microsoft's Link.exe."""
 
-    def __init__(self, for_machine: mesonlib.MachineChoice,
-                 *, version: str = 'unknown version'):
-        super().__init__(['link.exe'], for_machine, 'link', version=version)
+    def __init__(self, for_machine: mesonlib.MachineChoice, *,
+                 exelist: typing.List[str] = ['link.exe'], prefix: str = '',
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [B006] Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them. ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4487207))

<sub>posted by Sider</sub>

File: mesonbuild/environment.py [hunk 4]
```diff
@@ -1047,17 +1071,24 @@ def _detect_objc_or_objcpp_compiler(self, for_machine: MachineInfo, objc: bool)
...
             if 'clang' in out:
+                linker = None
                 comp = ClangObjCCompiler if objc else ClangObjCPPCompiler
                 if 'Apple' in out or self.machines[for_machine].is_darwin():
                     compiler_type = CompilerType.CLANG_OSX
                 elif 'windows' in out or self.machines[for_machine].is_windows():
                     compiler_type = CompilerType.CLANG_MINGW
+                    # If we're in a MINGW context this actually will use a gnu style ld
+                    try:
+                        linker = self._guess_win_linker(compiler, for_machine, cls.LINKER_PREFIX)
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__ [F821] undefined name 'cls' ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4487217))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 5]
```diff
@@ -763,33 +781,36 @@ def get_debug_crt_args(self) -> typing.List[str]:
...
         pdbarr = targetfile.split('.')[:-1]
         pdbarr += ['pdb']
-        return ['/DEBUG', '/PDB:' + '.'.join(pdbarr)]
+        return self._apply_prefix('/DEBUG') + self._apply_prefix('/PDB:' + '.'.join(pdbarr))
 
     def get_link_whole_for(self, args: typing.List[str]) -> typing.List[str]:
         # Only since VS2015
         args = mesonlib.listify(args)
-        return ['/WHOLEARCHIVE:' + x for x in args]
+        l = []  #typing.List[str]
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [E262] inline comment should start with '# ' ([link](https://sider.review/gh/repos/19784232/pulls/5856#issue-4491600))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
@jpakkane, can we merge this sooner and deal with the #5854 after? This is also blocking #5833.

**Comment by jpakkane** [comment]
Looking good.

**Pull request merged**